packages/website/blog/2024-08-15-parser-options-project-service.mdx

```diff
@@ -0,0 +1,260 @@
+> Manual Program creation logic comes with a few issues:
+>
+> - Complex project setups can be difficult to get right.
+>   - For example, [typescript-eslint does not yet support Project References](https://github.com/typescript-eslint/typescript-eslint/issues/2094).
+> - The TypeScript compiler options used in the user's editor might differ from the compiler options in the TSConfigs they specified on disk.
+> - Files not included in created Programs can't be linted with type information, even though editors still typically surface type information when editing those files.
+>   - Most commonly, `.eslintrc.(c)js` files can be tricky to lint, resulting in the dreaded [_TSConfig does not include this file_ error](/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file).
+>
+> We're working on an option to instead call the same TypeScript "Project Service" APIs that editors such as VS Code use to create Programs for us instead.
+> Project Services will automatically detect the TSConfig for each file (like `project: true`), and will also allow type information to be computed for JavaScript files without the `allowJs` compiler option (unlike `project: true`).
+
+Following a year of discussion and testing, we believe the new Project Service API is ready to be used by real-world projects.
+We've found it to be generally faster at runtime and more straightforward to configure.
+
+We're therefore promoting the `parserOptions.EXPERIMENTAL_useProjectService` option to the stable name **`parserOptions.projectService`** in typescript-eslint v8.
+
+## ✍️ Onboarding to the Project Service
+
+You can change over to the new Project Service API by replacing `project` with `projectService` in your ESLint configuration:
+
+<Tabs groupId="eslint-config">
+<TabItem value="Flat Config">
+
+```js title="eslint.config.js"
+export default tseslint.config({
+  // ...
+  languageOptions: {
+    parserOptions: {
+      // Remove this line
+      project: true,
+      // Add this line
+      projectService: true,
+      tsconfigRootDir: import.meta.dirname,
+    },
+  },
+  // ...
+});
+```
+
+:::note
+[`import.meta.dirname`](https://nodejs.org/api/esm.html#importmetadirname) is only present for ESM files in Node.js >=20.11.0 / >= 21.2.0.<br />
+For CommonJS modules and/or older versions of Node.js, [use `__dirname` or an alternative](https://stackoverflow.com/questions/46745014/alternative-for-dirname-in-node-js-when-using-es6-modules).
+:::
+
+</TabItem>
+<TabItem value="Legacy Config">
+
+```js title=".eslintrc.cjs"
+module.exports = {
+  // ...
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    // Remove this line
+    project: true,
+    // Add this line
+    projectService: true,
```

Comment:
If I have an ESLint configuration in a separate package, should I set `projectService` to `true` and let the linter find the nearest `tsconfig.json` in the user's environment, or I have to pass the tsconfig path from a user land, e.g. via `path.resolve("./tsconfig.json")`?

In `typescript-eslint@8.0.0-alpha.52`, I encounter this error when creating a new file:

> Parsing error: A:\projects\55-sanaye\src\logic\requests\new-file.ts was not found by the project service. Consider either including it in the `tsconfig.json` or including it in `allowDefaultProject`.

Even though the file is already included in `tsconfig.json`, I had to use <kbd>CTRL</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd> > `ESLint: Restart ESLint server` to resolve the error. Is this issue related to `projectService: true` or VS extension caching?

Reply:
👋 @AhmedBaset as excited as I am that you're trying out `projectService`, this isn't the right place to ask for help with it. See https://typescript-eslint.io/contributing/issues#questions-and-support-requests for the right places. They'll prompt you to provide the info we'd need to offer support.

You might also want to see our FAQs on https://typescript-eslint.io/troubleshooting/faqs, in particular https://typescript-eslint.io/troubleshooting/faqs/general#changes-to-one-file-are-not-reflected-when-linting-other-files-in-my-ide.